### ISSUE TITLE
Danger with packaging vscode-jsonrpc twice - Unknown parameter structure

### ISSUE THREAD

**Issue by Ferix9288** [TypeFox/monaco-languageclient#254]
tl;dr Remove dependency on vscode-jsonrpc in `example` dir because it causes things to break. See this issue https://github.com/microsoft/vscode-languageserver-node/pull/719

## Context & The Problem

So after I merged in @CGNonofr 's latest PR into my custom fork, my example all of sudden broke. Specifically, I was getting this trace error:

```
console-window.ts:11 Couldn't start client Sample Language Client
showMessage @ console-window.ts:11
showMessage @ vscode-api.ts:853
(anonymous) @ client.js:2168
Promise.then (async)
start @ client.js:2164
onConnection @ client.ts:227
webSocket.onopen @ connection.ts:19
connection.js:776 Uncaught (in promise) Error: Unknown parameter structure byName
    at computeSingleParam (connection.js:776)
    at computeMessageParams (connection.js:787)
    at Object.sendRequest (connection.js:929)
    at Object.initialize (connection.ts:107)
    at MonacoLanguageClient.doInitialize (client.js:2220)
    at MonacoLanguageClient.initialize (client.js:2216)
    at client.js:2163
computeSingleParam @ connection.js:776
computeMessageParams @ connection.js:787
sendRequest @ connection.js:929
initialize @ connection.ts:107
doInitialize @ client.js:2220
initialize @ client.js:2216
(anonymous) @ client.js:2163
Promise.then (async)
start @ client.js:2164
onConnection @ client.ts:227
webSocket.onopen @ connection.ts:19
```

When I debugged into the code, the culprit is that somehow byName kind is not recognized even though the payload looks clean.

```
parameterStructures: ParameterStructures
kind: "byName"
```

```js
    function computeSingleParam(parameterStructures, param) {
        switch (parameterStructures) {
            case messages_1.ParameterStructures.auto:
                if (isNamedParam(param)) {
                    return nullToUndefined(param);
                }
                else {
                    return [undefinedToNull(param)];
                }
                break;
            case messages_1.ParameterStructures.byName:
                if (!isNamedParam(param)) {
                    throw new Error(`Recevied parameters by name but param is not an object literal.`);
                }
                return nullToUndefined(param);
            case messages_1.ParameterStructures.byPosition:
                return [undefinedToNull(param)];
            default:
                throw new Error(`Unknown parameter structure ${parameterStructures.toString()}`);
        }
    }
```

I was extremely confused as Loic's changelog shouldn't impact this at all; in fact, I put a breakpoint at every diff and the new code never got executed. Turns out, the culprit is that `ParameterStructures` is meant to be a singleton but as  dbaeumer stated in the linked issue, having two packaged instances (one from monaco-languageclient and the other for the example) meant there were two instances. Seems like this is a ticking bomb waiting to happen, and it just so happened that re-compiling the code with the changes somehow exposed this error to occur deterministically for me.

## The Solution & Proposal

Remove vscode-jsonrpc in package.json of `example` and instead import `MessageConnection` from monaco-languageclient as this seems to be imported in connections.ts and then re-exported in index.ts.

Specifically, my diff looks like below (I can create a PR as well if that's the preference). After my change, it works again :D

```
-import {MessageConnection} from 'vscode-jsonrpc';
+// import {MessageConnection} from 'vscode-jsonrpc';
 import {
     MonacoLanguageClient,
     CloseAction,
@@ -12,6 +12,7 @@ import {
     MonacoServices,
     createConnection,
     Configurations,
+    MessageConnection,
     WorkspaceConfiguration,
     MonacoProvideCompletionItemsSignature,
 } from 'monaco-languageclient';

```

**Comment by CGNonofr** [comment]
Ok, thanks for the bug report!

I already had the exact same issue when debugging it using a local npm dependency. It's a known issue to have duplicated sub-dependencies when using local dependencies so I didn't really mind.

You can create a PR if you feel it, at least you'll get the credit for the fix.
Otherwise just tell me and I'll get a look at it.

**Comment by Ferix9288** [comment]
Much appreciated @CGNonofr  for the quick response! I created the PR but honestly don't mind if you need to end up doing it due to permission issues.

https://github.com/TypeFox/monaco-languageclient/pull/255

**Comment by Ferix9288** [comment]
> I already had the exact same issue when debugging it using a local npm dependency. It's a known issue to have duplicated sub-dependencies when using local dependencies so I didn't really mind.

How did you end up resolving it? I'm actually having some issues taking my example and porting it over to my company's application because of this issue. I do see that I have two instances of vscode-jsonrpc though, so I'm hoping the fix is similar.

**Comment by CGNonofr** [comment]
I just resolved the issue by not using local dependencies at all

To test it I just build it, package it, then install the package.

If you need local dependencies and you use webpack, there is a solution using aliases with an absolute path.

**Comment by mofux** [comment]
Thanks for the very detailed explanation and the PR. The PR is merged now.

**Comment by jchnxu** [comment]
I met this problem because there's 2 vscode-languageserver-protocol package. One in `node_modules/vscode-languageserver-protocol`, one in `another-package/node_modules/vscode-languageserver-protocol`.

I believe the protocol is exposed in the API hence should be declared as peerDependencies, not dependencies.
